File: Java/Inheritance/InterfaceTest2/Program2.java
import taxation.*;

class Auditor implements AutoCloseable{

	Auditor() {
		System.out.println("Acquiring departmental resources");
	}

	public void audit(TaxPayer person, String id) {
		if(id.length() < 4)
			throw new IllegalArgumentException("Invalid ID");
		System.out.printf("%s has to pay %.2f as income tax%n", id, person.incomeTax());
	}

	public void close() {
		System.out.println("Releasing departmental resources");
	}
}

class Program {

	private static void processDealer(double sales) {
		var d = new Dealer(sales);
		var a = new Auditor();
		a.audit(d, "Jack");
		a.close();
	}

	private static void processWorker(double wages, String name) {
		var w = new Worker(wages);
		//using try-with-resources to implicitly call close method of an AutoCloseable object
		try(var a = new Auditor()){
			a.audit(w, name);
		}		
	}

	public static void main(String[] args) {
		try{
			double earn = Double.parseDouble(args[0]);
			if(args.length < 2)
				processDealer(earn);
			else
				processWorker(earn, args[1]);
		}catch(Exception e){
			System.out.printf("Error: %s%n", e);
		}
	}
}

File: DotNet/Inheritance/InterfaceTest2/DemoApp/Program2.cs
using Taxation;

class Auditor : IDisposable
{
    public Auditor()
    {
        Console.WriteLine("Acquiring deparmental resources.");
    }

    public void Audit(string id, ITaxPayer payer)
    {
        if(id.Length < 4)
            throw new ArgumentException("Invalid ID");
        Console.WriteLine("{0} has to pay {1:0.00} as income tax", id, payer.IncomeTax() + 500);
    }

    public void Dispose()
    {
        Console.WriteLine("Releasing deparmental resources.");
    }
}

class Program
{
    private static void ProcessDealer(Dealer d)
    {
        var a = new Auditor();
        a.Audit("Jack", d); //dealer will be converted to ITaxPayer through boxing
        a.Dispose();
    }

    private static void ProcessWorker(Worker w, string name)
    {
        using(var a = new Auditor())
        {
            a.Audit(name, w);
        }

    }

    static void Main(string[] args)
    {
        try
        {
            decimal amount = decimal.Parse(args[0]);
            if(args.Length < 2)
            {
                Dealer d;
                d.TotalSales = amount;
                ProcessDealer(d);
            }
            else
            {
                Worker w;
                w.MonthlyWages = amount;
                ProcessWorker(w, args[1]);
            }
        }   
        catch(Exception ex)
        {
            Console.WriteLine("Error: {0}", ex.Message);
        }    
    }
}